perf/light.cc

```diff
@@ -0,0 +1,578 @@
+			}
+		}
+		state.SetItemsProcessed(processed);
+	}
+
+	T hash_table;
+};
+
+class LightAdapter {
+public:
+	LightAdapter()
+	{
+		light_create(&ht, light_extent_size, light_malloc_extend,
+			     light_free_extend, &extents_count, 0);
+	}
+	~LightAdapter()
+	{
+		light_destroy(&ht);
+	}
+
+	// Functions return smth in order to suppress "error: invalid use of void expression".
+	int
+	erase(const Tuple &tuple)
+	{
+		light_delete_value(&ht, tuple.hash, tuple.raw);
+		return 0;
+	}
+
+	int
+	insert(const Tuple &tuple)
+	{
+		light_insert(&ht, tuple.hash, tuple.raw);
+		return 0;
+	}
+
+	TupleRaw
+	find(const Tuple &tuple)
+	{
+		static TupleRaw DUMMY{nullptr, 0};
+		uint32_t slot = light_find(&ht, tuple.hash, tuple.raw);
+		if (slot != light_end)
+			return light_get(&ht, slot);
+		return DUMMY;
+	}
+
+	uint32_t
+	find_key(const Tuple &tuple)
+	{
+		return light_find_key(&ht, tuple.hash, tuple.key);
+	}
+
+	void
+	clear()
+	{
+		light_destroy(&ht);
+		light_create(&ht, light_extent_size, light_malloc_extend,
+			     light_free_extend, &extents_count, 0);
+	}
+
+	void
+	reserve(std::size_t) { /* no-op*/ }
+
+	std::size_t
+	iter_all()
+	{
+		std::size_t processed = 0;
+		struct light_iterator iter;
+		light_iterator_begin(&ht, &iter);
+		Value_t *p = nullptr;
+		while ((p = light_iterator_get_and_next(&ht, &iter)) != nullptr) {
+			benchmark::DoNotOptimize(p->data);
+			processed++;
+		}
+		return processed;
+	}
+private:
+	struct light_core ht;
+};
+
+using USet = std::unordered_set<Tuple, Hash::TupleHash, TupleEqual>;
+
+class STLAdapter {
+public:
+	auto erase(const Tuple &tuple) { return ht.erase(tuple); }
+	auto insert(const Tuple &tuple) { return ht.insert(tuple); }
+	auto find(const Tuple &tuple) { return ht.find(tuple); }
+	auto find_key(const Tuple &tuple) { return ht.find(tuple); }
+	void clear() { ht.clear(); }
+	void reserve(std::size_t n) { ht.reserve(n); }
+
+	std::size_t
+	iter_all()
+	{
+		std::size_t processed = 0;
+		for (const auto itr : ht) {
+			benchmark::DoNotOptimize(itr.raw);
+			processed++;
+		}
+		return processed;
+	}
+
+private:
+	USet ht;
+};
+
+#define BENCHMARK_TEMPLATE_REGISTER(METHOD_NAME, HASH_TABLE_NANE, STORAGE) \
+	BENCHMARK_TEMPLATE_DEFINE_F(HashTableFixture, HASH_TABLE_NANE##METHOD_NAME, STORAGE)(benchmark::State& state) \
+	{ HashTableFixture::METHOD_NAME(state); } \
+	BENCHMARK_REGISTER_F(HashTableFixture, HASH_TABLE_NANE##METHOD_NAME)-> \
+		RangeMultiplier(TUPLE_COUNT_MULTIPLIER)-> \
+			Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX)
+
+#define BENCHMARK_TEMPLATE_REGISTER_LIGHT(METHOD_NAME) \
+	BENCHMARK_TEMPLATE_REGISTER(METHOD_NAME, Light, LightAdapter)
+
+#define BENCHMARK_TEMPLATE_REGISTER_STL(METHOD_NAME) \
+	BENCHMARK_TEMPLATE_REGISTER(METHOD_NAME, STL, STLAdapter)
+
+#define BENCHMARK_TEMPLATE_REGISTER_FOR_ALL_IMPLS(METHOD_NAME) \
+	BENCHMARK_TEMPLATE_REGISTER_LIGHT(METHOD_NAME); \
+	BENCHMARK_TEMPLATE_REGISTER_STL(METHOD_NAME)
+
+BENCHMARK_TEMPLATE_REGISTER_FOR_ALL_IMPLS(InsertRandValue);
```

Comment:
It would be great if test names was shorter. I mean `HashTableFixture<LightAdapter>/LightInsertRandValue/1000000`. It does not fit my laptop screen now.